### PR TITLE
Calendars properly log when bad things happen

### LINKED ISSUE #1648: Calendars with dead source URLs should show this to the user
# Problem

Calendars can have URLs that are unreachable/unsuccessful (404/501/401) that causes them to move in to "error" state but that state is also used to catch every other error mode the calendar can encounter. 

## User story

As a partner/site admin I would like to see clearly when my calendars are pointing to "unsuccessful request" URLs so I can change the URL or delete the calendar.

As a developer/platform admin I would like to see which calendars are "unsuccessful request" so they don't show up in the list of error-ing calendars (it is not an error that comes from PlaceCal itself).

## Acceptance criteria

- [ ] as an end user: calendars with bad source URLs have a distinct and clearly worded message indicating that the URL is not working.
- [ ] as an admin: calendars that have bad source URLs don't show up on the list of calendars with errors (as it is not an internal problem).

## Implementation notes & questions

- [ ] Put new state on calendar
- [ ] Put new transition method and filter method
- [ ] In admin calendar edit/show implement "bad url" error box UX
- [ ] In the developer job page verify calendars in this state don't show up.

### PR DESCRIPTION
**Pull request by ivan-kocienski-gfsc**
Fixes  #1648 

Work done:

- define new state in calendar state machine (`bad_response`).
- implement a calendar method to flag calendar with `bad_response`.
  (+ some tests for this)
- created a new exception in the importer
  (moved exceptions into their own place)
- catching the `bad_response` exception and flagging the calendar
  (in the activejob handler)

- moving all the code that downloads a remote HTTP URL into one place in the importer (the parser base).
  (+ validating HTTP get was a success)
  (+ raising exception if something goes wrong)
  (+ some tests for this)
  (+ this took a long time as we had a major bug in our code that meant all our tests started failing)
- changing the way JSON is parsed so bad JSON is flagged properly
  (+ tests for this)
- be flagged for bad_response.
  (+ tests for this)
- tried validating ICAL sources
  (+ tests for this)

Not Done:

- XML parsing error detection/reporting to calendar
- leaving eventbrite untouched as it is too far out of scope for this issue.

### CODE REVIEW

File: app/jobs/calendar_importer/exceptions.rb
Comment (kimadactyl) [inline app/jobs/calendar_importer/exceptions.rb:14]: InvalidResponse maybe better name? 'Bad' I find a bit vague (why was it bad?)

**Comment by kimadactyl** [comment]
Small questions I can't quite see:

- Would it be good to have the proper tags for static code documentation showing what functions take and return?
- How are we storing the bad response code?

Otherwise looks great just would have preferred to break this into slightly smaller chunks but understand why you did this!

**Comment by ivan-kocienski-gfsc** [comment]
@kimadactyl at the moment the bad response code is stored in the error field and looks something like `The source URL could not be read (code=401)` (as defined [here](https://github.com/geeksforsocialchange/PlaceCal/blob/90145eef69ce884ba5e92c2a021f036b4aa71fc2/app/jobs/calendar_importer/parsers/base.rb#L70))

**Comment by katjam** [comment]
:cry: I will take a look but there are 48 files changed here!

File: app/jobs/calendar_importer/parsers/ics.rb
Comment (katjam) [inline app/jobs/calendar_importer/parsers/ics.rb:63]: How could we test or investigate that?

Comment (ivan-kocienski-gfsc) [reply app/jobs/calendar_importer/parsers/ics.rb:63]: i attempted to test this [here](https://github.com/geeksforsocialchange/PlaceCal/blob/ik-1648-calendar-bad-source-state/test/jobs/calendar_importer/event_parse_ics_test.rb).

but ultimately the icalendar gem appears to lack a distinct exception class.

as for investigating- i think you'd have to trace the code path manually and see how the ICS parser tests for errors. so if there is a problem it will likely be in that code (or our input into it, which is tested on the line above). at least it is not catching `StandardError`.

or we could fork the icalendar gem and give it proper error reporting.

Comment (kimadactyl) [reply app/jobs/calendar_importer/parsers/ics.rb:63]: maybe good to put all this in the comment for later?

**Review by katjam — COMMENTED** [review]
I've left a few comments. None of them dealbreakers.

I'm pretty sure this can go in as is, but holding off on clicking approve because wanted to make sure the use of VCR should be part of this. It read through about 10 tests from top and bottom - but it was hard to see how the changes there related to the centralising of the error handling.

That makes me feel like I am missing some context. Like when we moved the error handling for event url validation, did all the tests break/fail?

File: app/jobs/calendar_importer/parsers/ics.rb
Comment (ivan-kocienski-gfsc) [reply app/jobs/calendar_importer/parsers/ics.rb:74]: Hopefully. Have to keep an eye on the errors and hope we remember that this is what is happening. Half tempted to fork the ICS gem and put in some decent exception classes so it's easier to pull out ICS import problems

File: test/controllers/admin/calendars_controller_test.rb
Comment (ivan-kocienski-gfsc) [reply test/controllers/admin/calendars_controller_test.rb:7]: The problem is that in the calendar model there is [this](https://github.com/geeksforsocialchange/PlaceCal/blob/428a0a7351347aaf8e81b303752a9a40b25c3f73/app/models/calendar.rb#L98) code that is triggered before save (when source has changed value). This then instantiates a CalendarImporter object which causes it to run [this](https://github.com/geeksforsocialchange/PlaceCal/blob/428a0a7351347aaf8e81b303752a9a40b25c3f73/app/jobs/calendar_importer/calendar_importer.rb#L24) validation. But if you look down the code to [here](https://github.com/geeksforsocialchange/PlaceCal/blob/428a0a7351347aaf8e81b303752a9a40b25c3f73/app/jobs/calendar_importer/calendar_importer.rb#L51) it swallows all the exceptions. which is bad.

in order to fulfil the acceptance criteria i had to refactor this bit of code so it only happens in one place and it only tests for failure conditions that it is concerned with.

the pain was that when taking out this `rescue StandardError` our test suite went on fire. not because i had broken something but because now we weren't blanket suppressing exceptions any more all the bugs in our test suite suddenly lit up.

Comment (ivan-kocienski-gfsc) [reply test/controllers/admin/calendars_controller_test.rb:7]: (this also led to a massive problem with our test factories)

File: test/controllers/admin/collections_controller_test.rb
Comment (ivan-kocienski-gfsc) [reply test/controllers/admin/collections_controller_test.rb:8]: i don't think so. i have a tendancy to put blank lines between a lot of my code to break up the wall of text but sometimes they can be a bit redundant

**Review by kimadactyl — APPROVED** [review]
Thanks for review Katja - I'm happy with Ivan's replies so happy to merge when you are Ivan

**Comment by kimadactyl** [comment]
Does this resolve https://github.com/geeksforsocialchange/PlaceCal/issues/1642 ?

**Pull request merged**
